File: 이전_Pycharm_lym-master/algorithm_lym/A형 보충/A형보충_1_DFS/단지번호붙이기/단지번호붙이기.java
import java.io.FileInputStream;
import java.util.*;

public class Main
{
	static int N;   //정사각형이므로 가로와 세로의 크기는 같으며 5≤N≤25
	static int[][] map;
	static boolean[][] visited; 
	static int group;
	static int dx[] = {-1, 1, 0, 0};
	static int dy[] = {0, 0, -1, 1};
	static int count[];
	
	static void dfs(int x, int y, int group)
	{
		int nx, ny;
		visited[x][y] = true;
		map[x][y] = group;
		count[group]++;
		
		for(int i=0; i<4; i++) {
			nx = x + dx[i];
			ny = y + dy[i];
			
			if(nx < 0 || nx >= N || ny < 0 || ny >= N) continue;
			if(visited[nx][ny]) continue;
			if(map[nx][ny] == 0) continue;
			dfs(nx, ny, group);		
		}
	}
	static void printArr(int map[][])
	{
		for(int i=0; i < N; i++) {
			for(int j=0; j < N; j++) {
				System.out.print(map[i][j] + " ");
			}
			System.out.println();
		}
		System.out.println();
	}
	
	public static void main(String args[]) 
	{

		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		map = new int[N][N];
		visited = new boolean[N][N];
		group = 1;
		count = new int[N*N+1];
		String input;
		
		for(int i=0; i < N; i++) {
			input = sc.next();
			for(int j=0; j < N; j++) {
				map[i][j] = input.charAt(j) - '0';
			}
		}
		
		//printArr(map);
		
		for(int i=0; i < N; i++) {
			for(int j=0; j < N; j++) {
				if(map[i][j] != 0 && visited[i][j] == false)
				{
					dfs(i, j, group);
					group++;
				}
			}
		}
		//printArr(map);

		System.out.println(group-1);
		Arrays.sort(count);
		for(int i : count)
			if(i != 0)
				System.out.println(i);

		sc.close();
	}
}

File: SSAFY/이전_Pycharm_lym-master/algorithm_lym/A형 보충/A형보충_1_DFS/단지번호붙이기/단지번호붙이기.py
def dfs(x, y, group):
    visited[x][y] = 1;
    mat[x][y] = group;
    count[group] += 1;

    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]

        if not (0 <= nx < N and 0 <= ny < N): continue
        if visited[nx][ny]: continue
        if mat[nx][ny] == 0 : continue
        dfs(nx, ny, group)

import sys
sys.stdin = open("단지번호붙이기_input.txt")
N = int(input())
mat = [list(map(int, input())) for _ in range(N)] # 단지 입력 받음
visited = [[0 for _ in range(N)] for _ in range(N)] # 방문이력 들어갈 NxN 리스트 만들기
group = 1
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
count = [0] * (N*N+1) # +1 한 이유는 인덱스 0을 비워두고 사용하려고?

for i in range(N):
    for j in range(N):
        if mat[i][j] and visited[i][j] == 0: # mat !=0 and visited == 0 인 경우
            dfs(i, j, group)
            group += 1

count.sort()
print(group-1)
for i in count:
    if i != 0:
        print(i)
